### ISSUE TITLE
share/p2p/peers: Only add nodes to PeerManager that are compatible on desired protocols

### ISSUE THREAD

**Issue by renaynay** [celestiaorg/celestia-node#2604]
For lights, that's `shrex-nd` and for fulls, that's `shrex-nd` and `shrex-eds`

**Implementation**

Peer manager should look into peer store and check of peer supports required protocol. If peer does not support any protocols peer should not be added to peers pools.

**Comment by Wondertan** [comment]
Can be filtered via Peerstore which keep supported protocols information

**Comment by derrandz** [comment]
Proposed an initial PR, would be great having your feedback guys, thanks. 👍🏻

**Comment by walldiss** [comment]
@renaynay Could you please add more details on whats is the motivation for this feature? Asking, to clarify, if peer-manager should aggregate peers with ALL supported protocols or ANY? What if one of the peers implements only some of the protocols? Considering peer-manager is used to aggregate peers for multiple protocols, would it imply, that peer-manager interface should be changed, to allow getting peer by specific protocolID?
```
func (m *Manager) Peer(ctx context.Context, protocolID string, datahash share.DataHash)  (peer.ID, DoneFunc, error) 
```

**Comment by derrandz** [comment]
@walldiss Good point. If the choice is made to aggregate peers with ANY supported protocol, then this would have an impact on how we store such peers in the peer manager, and we should index them by their supported protocols, and thus (_as you mentioned_) would change the interface to support fetching peers by their specific protocol.ID

But I wonder if this might open room for some weird behavior, because if a full node is on the wrong protocol for `shrex-eds` but is on the right protocol for `shrex-nd`, I wonder if this would somehow break `shrex-nd`, cuz maybe the format of the retrieved shares from `shrex-eds` is out-dated and thus would (_dunno_) end up with breaking `shrex-nd` as a result?

Maybe the other question then is: do `shrex-eds` versions require specific `shrex-nd` versions for full nodes..?

Explaining the motivation for this feature request would help answer these questions properly.

**Comment by renaynay** [comment]
> @renaynay Could you please add more details on whats is the motivation for this feature? Asking, to clarify, if peer-manager should aggregate peers with ALL supported protocols or ANY? What if one of the peers implements only some of the protocols? Considering peer-manager is used to aggregate peers for multiple protocols, would it imply, that peer-manager interface should be changed, to allow getting peer by specific protocolID?

@walldiss Yes the peer manager should be able to grab the peer that best supports its need (for example, shrex-nd should choose peer that supports the current version of shrex-nd protocol that it needs)

Right now, our discovery adds any node that advertises on the `full` namespace and otherwise, adds peers from shrex-sub. That doesn't necessarily mean that those peers support the desired version of shrex-nd, for example

**Comment by walldiss** [comment]
**Implementation**

Peer manager should look into peer store and check of peer supports required protocol. If peer does not support any protocols peer should not be added to peers pools.

**Comment by loglapa** [comment]
I would like to take this after the RPC change. I will verify the current peer-manager architecture, filter discovered peers against the required protocol set through the peerstore, and add focused tests for compatible, partially compatible, and incompatible peers.
